java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -0,0 +1,476 @@
+        break;
+      case CATCH:
+        tree = ((CatchTree) st).catchKeyword();
+        break;
+      case BLOCK:
+        tree = ((BlockTree) st).openBraceToken();
+        if (st.parent().is(IF_STATEMENT)) {
+          tree = ((IfStatementTree) st.parent()).elseKeyword();
+        }
+        break;
+      default:
+        break;
+    }
+
+    if (nestingLevel > 0) {
+      flow.add(new JavaFileScannerContext.Location("+" + hit +" (incl " + nestingLevel + " for nesting)", tree));
+    } else {
+      flow.add(new JavaFileScannerContext.Location("+" + hit, tree));
+    }
+  }
+
+
+  private int countAnonymousClasses(Tree st, List<JavaFileScannerContext.Location> flow, int nestingLevel) {
+    int total = 0;
+
+    if (st.is(Tree.Kind.CLASS) && ! ((ClassTree)st).members().isEmpty()) {
+
+      for (Tree member : ((ClassTree) st).members()) {
+        if (member.is(Tree.Kind.METHOD) && ((MethodTree) member).block().body() != null) {
+          total += countComplexity(((MethodTree) member).block().body(), flow, nestingLevel+1);
+        }
+      }
+    }
+
+    return total;
+  }
+
+  private int countIfElseChains(Tree st, List<JavaFileScannerContext.Location> flow, int nestingLevel) {
+    int total = 0;
+
+    if (! st.is(Tree.Kind.IF_STATEMENT)) {
+      return total;
+    }
+
+    if (((IfStatementTree) st).elseStatement() == null) {
+      return total;
+    }
+
+    StatementTree elseStatement = ((IfStatementTree) st).elseStatement();
+
+    total++;
+
+    if (elseStatement.is(Tree.Kind.IF_STATEMENT)) {
+      addSecondaryLocation(flow, elseStatement, total, nestingLevel);
+
+      total += countConditions(((IfStatementTree) elseStatement).condition(), flow);
+      total += countIfElseChains(elseStatement, flow, nestingLevel);
+    } else {
+      // `else`
+      total += nestingLevel;
+      addSecondaryLocation(flow, elseStatement, total, nestingLevel);
+    }
+
+    List<StatementTree> children = getChildren(elseStatement);
+    total += countComplexity(children, flow, nestingLevel+1);
+
+    return total;
+  }
+
+  private int countExtraConditions(Tree tree, List<JavaFileScannerContext.Location> flow) {
+    switch (tree.kind()) {
+      case WHILE_STATEMENT:
+        return countConditions(((WhileStatementTree) tree).condition(), flow);
+      case FOR_STATEMENT:
+        return countConditions(((ForStatementTree) tree).condition(), flow);
+      case DO_STATEMENT:
+        return countConditions(((DoWhileStatementTree) tree).condition(), flow);
+      case IF_STATEMENT:
+        return countConditions(((IfStatementTree) tree).condition(), flow);
+      default:
+        return 0;
+    }
+  }
+
+  /**
+   * Increment for each non-like operator. So:
+   *   a           // +0
+   *   a && b      // +1
+   *   a && b && c // +1
+   *   a && b || c // +2
+   *
+   * @param expressionTree
+   * @return
+   */
+  private int countConditions(ExpressionTree expressionTree, List<JavaFileScannerContext.Location> flow) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "countConditions" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)